File: AI-Reply-Android/app/src/main/java/kz/yerek/aireply/ai/AIReplyError.kt
package kz.yerek.aireply.ai

/**
 * Every way generating a reply can fail, as a value the UI localizes.
 *
 * Deliberately closed and deliberately coarse. The user needs to know what to
 * do next; they do not need an HTTP status, a JSON body or an upstream error
 * string, and showing them one would leak detail that has no business on a
 * keyboard above WhatsApp.
 */
sealed interface AIReplyError {

    /** Nothing was copied, or the copied value held no text. */
    data object NoSourceMessage : AIReplyError

    /** The copied message is longer than the limit. */
    data class MessageTooLong(val limit: Int) : AIReplyError

    /**
     * The clipboard could not be read. Reachable on iOS when Full Access is off;
     * on Android it means the clip was empty, unreadable, or flagged sensitive
     * by the app that produced it.
     */
    data object ClipboardUnavailable : AIReplyError

    /** No API key entered (direct mode) or no service URL set (backend mode). */
    data object NotConfigured : AIReplyError

    /** Device is offline. */
    data object Offline : AIReplyError

    /** The request exceeded its time budget. */
    data object TimedOut : AIReplyError

    /** The user or the keyboard cancelled before a reply arrived. */
    data object Cancelled : AIReplyError

    /** Credential rejected: the key is wrong, revoked or out of quota. */
    data object AuthenticationFailed : AIReplyError

    /** Too many requests, ours or the provider's. */
    data object RateLimited : AIReplyError

    /** The service answered, but not with a usable reply. */
    data object EmptyResponse : AIReplyError

    /** Anything else: a 5xx, a malformed payload, an unreachable host. */
    data object ServiceUnavailable : AIReplyError
}

/** Thrown across suspend boundaries; the payload is what the UI actually reads. */
class AIReplyException(val error: AIReplyError) : Exception(error::class.simpleName)

fun AIReplyError.raise(): Nothing = throw AIReplyException(this)

File: AI-Reply/Shared/AI/AIReplyError.swift
import Foundation

/// Every way generating a reply can fail, as a value the UI localizes.
///
/// Deliberately closed and deliberately coarse. The user needs to know what to
/// do next; they do not need an HTTP status, a JSON body or an upstream error
/// string, and showing them one would leak detail that has no business on a
/// keyboard above WhatsApp.
enum AIReplyError: Error, Equatable, Sendable {
    /// Nothing was copied, or the copied value held no text.
    case noSourceMessage
    /// The copied message is longer than the 300-character limit.
    case messageTooLong(limit: Int)
    /// The keyboard cannot read the clipboard without Full Access.
    case fullAccessRequired
    /// The account is not ready to make a request.
    case notConfigured
    /// Device is offline.
    case offline
    /// The request exceeded its time budget.
    case timedOut
    /// The user or the keyboard cancelled before a reply arrived.
    case cancelled
    /// The account session is missing, expired or rejected.
    case authenticationFailed
    /// The account limit or request rate limit has been reached.
    case rateLimited
    /// The service answered, but not with a usable reply.
    case emptyResponse
    /// Anything else: a 5xx, a malformed payload, an unreachable host.
    case serviceUnavailable
}
